Userland/Libraries/LibWeb/Bindings/WindowProxy.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2022, Linus Groh <linusg@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Forward.h>
+#include <LibJS/Forward.h>
+#include <LibJS/Runtime/Object.h>
+#include <LibWeb/Forward.h>
+
+namespace Web::Bindings {
+
+class WindowProxy final : public JS::Object {
+    JS_OBJECT(WindowProxy, JS::Object);
+
+public:
+    WindowProxy(JS::GlobalObject&, WindowObject&);
+    virtual ~WindowProxy() override = default;
+
+    virtual JS::ThrowCompletionOr<JS::Object*> internal_get_prototype_of() const override;
+    virtual JS::ThrowCompletionOr<bool> internal_set_prototype_of(Object* prototype) override;
+    virtual JS::ThrowCompletionOr<bool> internal_is_extensible() const override;
+    virtual JS::ThrowCompletionOr<bool> internal_prevent_extensions() override;
+    virtual JS::ThrowCompletionOr<Optional<JS::PropertyDescriptor>> internal_get_own_property(JS::PropertyKey const&) const override;
+    virtual JS::ThrowCompletionOr<bool> internal_define_own_property(JS::PropertyKey const&, JS::PropertyDescriptor const&) override;
+    virtual JS::ThrowCompletionOr<JS::Value> internal_get(JS::PropertyKey const&, JS::Value receiver) const override;
+    virtual JS::ThrowCompletionOr<bool> internal_set(JS::PropertyKey const&, JS::Value value, JS::Value receiver) override;
+    virtual JS::ThrowCompletionOr<bool> internal_delete(JS::PropertyKey const&) override;
+    virtual JS::ThrowCompletionOr<JS::MarkedVector<JS::Value>> internal_own_property_keys() const override;
+
+    WindowObject& window() { return m_window; }
+    WindowObject const& window() const { return m_window; }
+
+    // NOTE: Someone will have to replace the wrapped window object as well:
+    // "When the browsing context is navigated, the Window object wrapped by the browsing context's associated WindowProxy object is changed."
+    // I haven't found where that actually happens yet. Make sure to use a Badge<T> guarded setter.
```

Comment:
Given this, `m_window` will eventually become a pointer (not a reference.)
But this is ok for now.

Reply:
Ah, I had it as a `NonnullRefPtr<Window>` before realizing it actually needs a `WindowObject`, and replaced it with a reference without thinking about it too much. Will update now :)